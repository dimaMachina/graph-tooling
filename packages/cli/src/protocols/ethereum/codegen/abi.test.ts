import path from 'path';
import fs from 'fs-extra';
import { afterAll, beforeAll, describe, expect, test } from 'vitest';
import ABI from '../abi';
import AbiCodeGenerator from './abi';

let tempdir: string;
let abi: ABI;
let generatedTypes: any[];

describe.concurrent('ABI code generation', () => {
  beforeAll(async () => {
    tempdir = await fs.mkdtemp('abi-codegen');

    try {
      const filename = path.join(tempdir, 'ABI.json');
      await fs.writeFile(
        filename,
        JSON.stringify([
          {
            constant: true,
            inputs: [],
            name: 'read',
            outputs: [{ name: '', type: 'bytes32' }],
            payable: false,
            type: 'function',
          },
          {
            constant: true,
            inputs: [
              {
                name: 'proposalId',
                type: 'uint256',
              },
              {
                components: [
                  {
                    name: 'foo',
                    type: 'uint8',
                  },
                  {
                    name: 'bar',
                    type: 'tuple',
                    components: [{ name: 'baz', type: 'address' }],
                  },
                ],
                name: '',
                type: 'tuple',
              },
            ],
            name: 'getProposal',
            outputs: [
              {
                components: [
                  {
                    name: 'result',
                    type: 'uint8',
                  },
                  {
                    name: 'target',
                    type: 'address',
                  },
                  {
                    name: 'data',
                    type: 'bytes',
                  },
                  {
                    name: 'proposer',
                    type: 'address',
                  },
                  {
                    name: 'feeRecipient',
                    type: 'address',
                  },
                  {
                    name: 'fee',
                    type: 'uint256',
                  },
                  {
                    name: 'startTime',
                    type: 'uint256',
                  },
                  {
                    name: 'yesCount',
                    type: 'uint256',
                  },
                  {
                    name: 'noCount',
                    type: 'uint256',
                  },
                ],
                name: '',
                type: 'tuple',
              },
            ],
            payable: false,
            stateMutability: 'view',
            type: 'function',
          },
          {
            type: 'function',
            stateMutability: 'view',
            payable: 'false',
            name: 'getProposals',
            outputs: [
              {
                type: 'uint256',
                name: 'size',
              },
              {
                type: 'tuple[]',
                components: [
                  { name: 'first', type: 'uint256' },
                  { name: 'second', type: 'string' },
                ],
              },
            ],
          },
          {
            type: 'function',
            stateMutability: 'view',
            name: 'overloaded',
            inputs: [
              {
                type: 'string',
              },
            ],
            outputs: [
              {
                type: 'string',
              },
            ],
          },
          {
            type: 'function',
            stateMutability: 'view',
            name: 'overloaded',
            inputs: [
              {
                type: 'uint256',
              },
            ],
            outputs: [
              {
                type: 'string',
              },
            ],
          },
          {
            type: 'function',
            stateMutability: 'view',
            name: 'overloaded',
            inputs: [
              {
                type: 'bytes32',
              },
            ],
            outputs: [
              {
                type: 'string',
              },
            ],
          },
        ]),
        'utf-8',
      );
      abi = ABI.load('Contract', filename);
      const codegen = new AbiCodeGenerator(abi);
      generatedTypes = codegen.generateTypes();
    } finally {
      await fs.remove(tempdir);
    }
  });

  afterAll(async () => {
    await fs.remove(tempdir);
  });

  describe('Generated types', () => {
    test('All expected types are generated', () => {
      expect(generatedTypes.map(type => type.name)).toMatchSnapshot()
    });
  });

  describe('Contract class', () => {
    test('Exists', () => {
      expect(generatedTypes.find(type => type.name === 'Contract')).toBeDefined();
    });

    test('Has methods', () => {
      const contract = generatedTypes.find(type => type.name === 'Contract');
      expect(contract.methods).toBeInstanceOf(Array);
    });

    test('Has `bind` method', () => {
      const contract = generatedTypes.find(type => type.name === 'Contract');
      expect(contract.methods.find((method: any) => method.name === 'bind')).toBeDefined();
    });

    test('Has methods for all callable functions', () => {
      const contract = generatedTypes.find(type => type.name === 'Contract');
      expect(contract.methods.map((method: any) => method.name)).toContain('getProposal');
    });
  });

  describe('Methods for callable functions', () => {
    test('Have correct parameters', () => {
      const contract = generatedTypes.find(type => type.name === 'Contract');
      expect(contract.methods.map((method: any) => [method.name, method.params])).toMatchSnapshot()
    });

    test('Have correct return types', () => {
      const contract = generatedTypes.find(type => type.name === 'Contract');
      expect(contract.methods.map((method: any) => [method.name, method.returnType])).toMatchSnapshot()
    });
  });

  describe('Tuples', () => {
    test('Tuple types exist for function parameters', () => {
      let tupleType = generatedTypes.find(
        type => type.name === 'Contract__getProposalInputParam1Struct',
      );

      // Verify that the tuple type has methods
      expect(tupleType.methods).toBeDefined();

      // Verify that the tuple type has getters for all tuple fields with
      // the right return types
      expect(tupleType.methods.map((method: any) => [method.name, method.returnType])).toMatchSnapshot()

      // Inner tuple:
      tupleType = generatedTypes.find(
        type => type.name === 'Contract__getProposalInputParam1BarStruct',
      );

      // Verify that the tuple type has methods
      expect(tupleType.methods).toBeDefined();

      // Verify that the tuple type has getters for all tuple fields with
      // the right return types
      expect(tupleType.methods.map((method: any) => [method.name, method.returnType])).toMatchSnapshot()
    });

    test('Tuple types exist for function return values', () => {
      const tupleType = generatedTypes.find(
        type => type.name === 'Contract__getProposalResultValue0Struct',
      );

      // Verify that the tuple type has methods
      expect(tupleType.methods).toBeDefined();

      // Verify that the tuple type has getters for all tuple fields with
      // the right return types
      expect(tupleType.methods.map((method: any) => [method.name, method.returnType])).toMatchSnapshot()
    });

    test('Function bodies are generated correctly for tuple arrays', () => {
      const contract = generatedTypes.find(type => type.name === 'Contract');
      const getter = contract.methods.find((method: any) => method.name === 'getProposals');

      expect(getter.body).not.toContain('toTupleArray<undefined>');
      expect(getter.body).toContain(
        'result[1].toTupleArray<Contract__getProposalsResultValue1Struct>()',
      );
    });
  });
});
