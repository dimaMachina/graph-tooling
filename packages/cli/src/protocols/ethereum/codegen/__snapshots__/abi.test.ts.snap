// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ABI code generation > Generated types > All expected types are generated 1`] = `
[
  "Contract__getProposalResultValue0Struct",
  "Contract__getProposalInputParam1Struct",
  "Contract__getProposalInputParam1BarStruct",
  "Contract__getProposalsResultValue1Struct",
  "Contract__getProposalsResult",
  "Contract",
]
`;

exports[`ABI code generation > Methods for callable functions > Have correct return types 1`] = `
[
  [
    "bind",
    Immutable.List [
      Param {
        "name": "address",
        "type": "Address",
      },
    ],
  ],
  [
    "read",
    Immutable.List [],
  ],
  [
    "try_read",
    Immutable.List [],
  ],
  [
    "getProposal",
    Immutable.List [
      Param {
        "name": "proposalId",
        "type": "BigInt",
      },
      Param {
        "name": "param1",
        "type": "Contract__getProposalInputParam1Struct",
      },
    ],
  ],
  [
    "try_getProposal",
    Immutable.List [
      Param {
        "name": "proposalId",
        "type": "BigInt",
      },
      Param {
        "name": "param1",
        "type": "Contract__getProposalInputParam1Struct",
      },
    ],
  ],
  [
    "getProposals",
    Immutable.List [],
  ],
  [
    "try_getProposals",
    Immutable.List [],
  ],
  [
    "overloaded",
    Immutable.List [
      Param {
        "name": "param0",
        "type": "string",
      },
    ],
  ],
  [
    "try_overloaded",
    Immutable.List [
      Param {
        "name": "param0",
        "type": "string",
      },
    ],
  ],
  [
    "overloaded1",
    Immutable.List [
      Param {
        "name": "param0",
        "type": "BigInt",
      },
    ],
  ],
  [
    "try_overloaded1",
    Immutable.List [
      Param {
        "name": "param0",
        "type": "BigInt",
      },
    ],
  ],
  [
    "overloaded2",
    Immutable.List [
      Param {
        "name": "param0",
        "type": "Bytes",
      },
    ],
  ],
  [
    "try_overloaded2",
    Immutable.List [
      Param {
        "name": "param0",
        "type": "Bytes",
      },
    ],
  ],
]
`;

exports[`ABI code generation > Methods for callable functions > Have correct return types 2`] = `
[
  [
    "bind",
    NamedType {
      "name": "Contract",
    },
  ],
  [
    "read",
    NamedType {
      "name": "Bytes",
    },
  ],
  [
    "try_read",
    "ethereum.CallResult<Bytes>",
  ],
  [
    "getProposal",
    NamedType {
      "name": "Contract__getProposalResultValue0Struct",
    },
  ],
  [
    "try_getProposal",
    "ethereum.CallResult<Contract__getProposalResultValue0Struct>",
  ],
  [
    "getProposals",
    NamedType {
      "name": "Contract__getProposalsResult",
    },
  ],
  [
    "try_getProposals",
    "ethereum.CallResult<Contract__getProposalsResult>",
  ],
  [
    "overloaded",
    NamedType {
      "name": "string",
    },
  ],
  [
    "try_overloaded",
    "ethereum.CallResult<string>",
  ],
  [
    "overloaded1",
    NamedType {
      "name": "string",
    },
  ],
  [
    "try_overloaded1",
    "ethereum.CallResult<string>",
  ],
  [
    "overloaded2",
    NamedType {
      "name": "string",
    },
  ],
  [
    "try_overloaded2",
    "ethereum.CallResult<string>",
  ],
]
`;

exports[`ABI code generation > Tuples > Function bodies are generated correctly for tuple arrays 1`] = `
[
  [
    "get foo",
    "i32",
  ],
  [
    "get bar",
    "Contract__getProposalInputParam1BarStruct",
  ],
]
`;

exports[`ABI code generation > Tuples > Function bodies are generated correctly for tuple arrays 2`] = `
[
  [
    "get baz",
    "Address",
  ],
]
`;

exports[`ABI code generation > Tuples > Function bodies are generated correctly for tuple arrays 3`] = `
[
  [
    "get result",
    "i32",
  ],
  [
    "get target",
    "Address",
  ],
  [
    "get data",
    "Bytes",
  ],
  [
    "get proposer",
    "Address",
  ],
  [
    "get feeRecipient",
    "Address",
  ],
  [
    "get fee",
    "BigInt",
  ],
  [
    "get startTime",
    "BigInt",
  ],
  [
    "get yesCount",
    "BigInt",
  ],
  [
    "get noCount",
    "BigInt",
  ],
]
`;
